README.md
<p align=center>
  <a href="https://github.com/TimoKunze/ProgressBarControl/releases/tag/2.6.3">
    <img alt="Release 2.6.3 Release" src="https://img.shields.io/badge/release-2.6.3-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ProgressBarControl/releases">
    <img alt="Download ProgressBarControl" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/ProgressBarControl/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=ProgressBarControl&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# ProgressBarControl
An ActiveX control for Visual Basic 6 that can be used to display a progress.

I've developed this ActiveX control in 2003 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 2.6.3 - 31/01/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 2.6.2 - 11/01/2018

- Project published on GitHub.



Version 2.6.1 - 28/01/2016

- FIX: The control could crash during OLE drag'n'drop.
- FIX: The control could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 2.6.0 - 20/01/2016

- Added the IProgressBar::FinishOLEDragDrop method. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 2.5.0 - 10/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Reactivated data-binding capabilities. When the library is registered, it will call regtlib.exe msdatsrc.tlb before registering itself. This seems to fix the mess caused by the broken Windows Update from the beginning of 2013.
- Initial support for Windows 8.1.



Version 2.4.2 - 27/03/2013

- FIX: The ResizedControlWindow event was not fired when it should have been fired.
- Disabled data-binding capabilities as they cause "Object library not registered" errors when adding the control library to the project. This problem might be related to a recent Windows Update.
- Initial support for Windows 8.



Version 2.4.1 - 29/05/2012

- FIX: Clicking the control no longer raises the Validate event of the currently focused control.



Version 2.4.0 - 16/03/2012

- Added the IProgressBar::DisplayProgressInTaskBar property.
- Added the IProgressBar::hApplicationWindow property.



Version 2.3.0 - 13/11/2011

- Added the IProgressBar::DetectDoubleClicks property to completely deactivate double clicks and handle them as two single clicks instead.
- Added the IProgressBar::DisplayText property.
- Added the IProgressBar::Font property.
- Added the IProgressBar::ForeColor property.
- Added the IProgressBar::HAlignment property.
- Added the IProgressBar::RightToLeft property and declared the IProgressBar::RightToLeftLayout property deprecated.
- Added the IProgressBar::Text property.
- Added the IProgressBar::TextShadowColor property.
- Added the IProgressBar::TextShadowOffsetX property.
- Added the IProgressBar::TextShadowOffsetY property.
- Added the IProgressBar::UseSystemFont property.



Version 2.2.1 - 21/05/2011

- Fixed a bug that could make other, invisible windows visible again under rather complex circumstances. I'm not sure this could happen practically, but in theory it could happen and practically it did happen for the Frame control which had the same bug.
- FIX: The ResizedControlWindow event was fired when actually nothing had changed.



Version 2.2.0 - 08/01/2011

- Switched to Visual C++ 2010 and dropped support for Windows 2000.



Version 2.1.4 - 07/11/2010

- FIX: IOLEDataObject::SetDropDescription could cause crashes in other programs.



Version 2.1.3 - 02/01/2010

- Fixed a potential problem with persistance.
- Improved support for Aero drag images.
- Improved performance.
- FIX: The 'shift' parameter that many events have, often missed the ALT key.
- Deploy (stripped) symbols with the binaries to make debugging easier. This are the pdb files in case you wonder.



Version 2.1.2 - 11/07/2009

- Initial support for Windows 7.
- Extended the AppearanceConstants enumeration by 'aDefault'.
- Extended the DropDescriptionIconConstants enumeration by 'ddiUseDefault'.
- FIX: Drop descriptions didn't work as expected if the drag source was a 64 bit app.
- Massively reduced the usage of GDI handles.
- Made some internal changes that will make it easier for me to keep binary compatibility in future.



Version 2.1.1 - 30/10/2008

- FIX: Ignore calls to IOLEDataObject::SetDropDescription() if the drag image is in classic style.
- Fixed some GDI handle leaks.
- Fixed some drawing glitches.
- Slightly changed handling of the situation where the user presses a mouse button, drags the mouse out of the control's client area and releases the button.
- Removed the data-binding stuff (it was probably broken anyway).
- Switched to Visual C++ 2008 and dropped support for Windows 9x, Me and NT 4.0.
- Improved the documentation.



Version 2.1.0 - 04/12/2007

- Added the IOLEDataObject::GetDropDescription() method.
- Added the IOLEDataObject::SetDropDescription() method.
- FIX: Changes were not applied implicitly when jumping between multiple instances of the control with the property pages opened.
- Improved the documentation.



Version 2.0.2 - 27/02/2007

- FIX: The IProgressBar::CurrentValue property wasn't restored when changing the IProgressBar::BarStyle property with version 6.0 of comctl32.dll.
- Decreased the binaries' size.
- Improved the documentation.
- Ported the WTL samples to WTL 8.0.
- Made the samples DPI aware.
- Linked the new bug tracking system.



Version 2.0.1 - 21/09/2006

- FIX: The "MouseIcon" property didn't work as expected.
- FIX: The "Visible" property didn't work as expected.
- Improved setup.



Version 2.0.0 - 06/07/2006

- First public release.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// IProgressBar
// properties
#define DISPID_PROGBAR_ACTIVATEMARQUEE							1
#define DISPID_PROGBAR_APPEARANCE										2
#define DISPID_PROGBAR_BACKCOLOR										3
#define DISPID_PROGBAR_BARCOLOR											4
#define DISPID_PROGBAR_BARSTYLE											5
#define DISPID_PROGBAR_BORDERSTYLE									6
#define DISPID_PROGBAR_CURRENTVALUE									DISPID_VALUE
#define DISPID_PROGBAR_DISABLEDEVENTS								7
#define DISPID_PROGBAR_DONTREDRAW										8
#define DISPID_PROGBAR_ENABLED											DISPID_ENABLED
#define DISPID_PROGBAR_HOVERTIME										9
#define DISPID_PROGBAR_HWND													10
#define DISPID_PROGBAR_MARQUEESTEPDURATION					11
#define DISPID_PROGBAR_MAXIMUM											12
#define DISPID_PROGBAR_MINIMUM											13
#define DISPID_PROGBAR_MOUSEICON										14
#define DISPID_PROGBAR_MOUSEPOINTER									15
#define DISPID_PROGBAR_ORIENTATION									16
#define DISPID_PROGBAR_PROGRESSSTATE								17
#define DISPID_PROGBAR_REGISTERFOROLEDRAGDROP				18
#define DISPID_PROGBAR_RIGHTTOLEFTLAYOUT						19
#define DISPID_PROGBAR_SMOOTHREVERSE								20
#define DISPID_PROGBAR_STEPWIDTH										21
#define DISPID_PROGBAR_SUPPORTOLEDRAGIMAGES					22
#define DISPID_PROGBAR_VERSION											23
#define DISPID_PROGBAR_DETECTDOUBLECLICKS						27
#define DISPID_PROGBAR_DISPLAYTEXT									28
#define DISPID_PROGBAR_FONT													29
#define DISPID_PROGBAR_FORECOLOR										30
#define DISPID_PROGBAR_HALIGNMENT										31
#define DISPID_PROGBAR_RIGHTTOLEFT									32
#define DISPID_PROGBAR_TEXT													33
#define DISPID_PROGBAR_TEXTSHADOWCOLOR							34
#define DISPID_PROGBAR_TEXTSHADOWOFFSETX						35
#define DISPID_PROGBAR_TEXTSHADOWOFFSETY						36
#define DISPID_PROGBAR_USESYSTEMFONT								37
#define DISPID_PROGBAR_DISPLAYPROGRESSINTASKBAR			38
#define DISPID_PROGBAR_HAPPLICATIONWINDOW						39
// fingerprint
#define DISPID_PROGBAR_APPID												500
#define DISPID_PROGBAR_APPNAME											501
#define DISPID_PROGBAR_APPSHORTNAME									502
#define DISPID_PROGBAR_BUILD												503
#define DISPID_PROGBAR_CHARSET											504
#define DISPID_PROGBAR_ISRELEASE										505
#define DISPID_PROGBAR_PROGRAMMER										506
#define DISPID_PROGBAR_TESTER												507
// methods
#define DISPID_PROGBAR_ABOUT												DISPID_ABOUTBOX
#define DISPID_PROGBAR_CHANGECURRENTVALUE						24
#define DISPID_PROGBAR_LOADSETTINGSFROMFILE					25
#define DISPID_PROGBAR_REFRESH											DISPID_REFRESH
#define DISPID_PROGBAR_SAVESETTINGSTOFILE						26
#define DISPID_PROGBAR_FINISHOLEDRAGDROP						40


// _IProgressBarEvents
// methods
#define DISPID_PROGBARE_CLICK												DISPID_VALUE
#define DISPID_PROGBARE_CONTEXTMENU									1
#define DISPID_PROGBARE_DBLCLICK										2
#define DISPID_PROGBARE_DESTROYEDCONTROLWINDOW			3
#define DISPID_PROGBARE_MCLICK											4
#define DISPID_PROGBARE_MDBLCLICK										5
#define DISPID_PROGBARE_MOUSEDOWN										6
#define DISPID_PROGBARE_MOUSEENTER									7
#define DISPID_PROGBARE_MOUSEHOVER									8
#define DISPID_PROGBARE_MOUSELEAVE									9
#define DISPID_PROGBARE_MOUSEMOVE										10
#define DISPID_PROGBARE_MOUSEUP											11
#define DISPID_PROGBARE_OLEDRAGDROP									12
#define DISPID_PROGBARE_OLEDRAGENTER								13
#define DISPID_PROGBARE_OLEDRAGLEAVE								14
#define DISPID_PROGBARE_OLEDRAGMOUSEMOVE						15
#define DISPID_PROGBARE_RCLICK											16
#define DISPID_PROGBARE_RDBLCLICK										17
#define DISPID_PROGBARE_RECREATEDCONTROLWINDOW			18
#define DISPID_PROGBARE_RESIZEDCONTROLWINDOW				19
#define DISPID_PROGBARE_XCLICK											20
#define DISPID_PROGBARE_XDBLCLICK										21


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR														1
#define DISPID_ODO_GETCANONICALFORMAT								2
#define DISPID_ODO_GETDATA													3
#define DISPID_ODO_GETFORMAT												4
#define DISPID_ODO_SETDATA													5
#define DISPID_ODO_GETDROPDESCRIPTION								6
#define DISPID_ODO_SETDROPDESCRIPTION								7
